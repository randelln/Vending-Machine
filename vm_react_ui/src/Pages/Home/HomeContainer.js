import React, { useEffect, useState } from 'react';
import HomeView from './HomeView';
import { API_BASE_URL } from '../../constants';

//Makes API request, init - To know when the request is made the first time.

// setStateData - Will set the data that comes back from the API, sets the response to the state
//setDataReloading - Reloading the entire application
// setDataLoading -  Changes variable that is DataLoading
export const fetchData = (init, url, setStateData, setDataReloading, setDataLoading) => {
    init === true ? setDataReloading(true) : setDataLoading(true);
    fetch(url)
        .then(res => res.json())
        .then(response => {
            setTimeout(function(){init === true ? setDataReloading(false) : setDataLoading(false);
                setStateData(response)},3000)      
        })
        .catch(err => {
            window.alert(`An error has occurred ${err}`);
        });
};

// Updates the running total and decreases amount of coins available
// .map - For all coins inside the array, if this coin value equals my coin, decrease the coin amount
export const addCoins = (coinValue, setRunningTotal, coins, runningTotal, setCoins) => {
    setRunningTotal(runningTotal + coinValue);
    const updatedCoins = coins.map(coin =>
        coinValue === coin.value
            ? {
                name: coin.name,
                value: coin.value,
                amount: coin.amount - 1
            }
            : coin
    );
    //Gives back new array of coins after purchase
    setCoins(updatedCoins);
};
// Home Component - Creating state variables for rerendering
// States variables - Has the variable and the function to set the value. When function is called page will rerender
const HomeContainer = () => {
    const [products, setProducts] = useState([]);
    const [coins, setCoins] = useState([]);
    const [loadingProducts, setLoadingProducts] = useState(false);
    const [loadingCoins, setLoadingCoins] = useState(false);
    const [reLoadingProducts, setReLoadingProducts] = useState(false);
    const [reLoadingCoins, setReLoadingCoins] = useState(false);
    const [loadingBuy, setLoadingBuy] = useState(false);
    const [runningTotal, setRunningTotal] = useState(0);
// Needed both loading and reloading products for purpose of reloading sections of page
    const loadData = () => {
        fetchData(true, `${API_BASE_URL}/coins`, setCoins, setReLoadingCoins, setLoadingCoins);
        fetchData(true, `${API_BASE_URL}/products`, setProducts, setReLoadingProducts, setLoadingProducts);
        setRunningTotal(0)
    };
// LifeCycle method, first method that runs when app starts, normally used for API endpoint calls and fetching data from DB
    useEffect(() => {
        loadData();
    }, []);
//Passing down all props and functions to HomeView
// Homeview is a component, that accepts 2 props functions and state, which are objects
    return (
        <HomeView
            functions={{
                loadData,
                fetchCoins: () => {
                    fetchData(false, `${API_BASE_URL}/coins`, setCoins, setReLoadingCoins, setLoadingCoins);
                },
                fetchProducts: () => {
                    fetchData(false, `${API_BASE_URL}/products`, setProducts, setReLoadingProducts, setLoadingProducts);
                },
                addCoins: coinValue => {
                    addCoins(coinValue, setRunningTotal, coins, runningTotal, setCoins);
                }
            }}
            state={{
                products,
                coins,
                loadingProducts,
                loadingCoins,
                reLoadingProducts,
                reLoadingCoins,
                loadingBuy,
                setLoadingBuy,
                runningTotal,
                setRunningTotal
            }}
        />
    );
};

export default HomeContainer;
