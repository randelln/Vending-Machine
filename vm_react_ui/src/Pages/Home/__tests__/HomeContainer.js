import React from 'react';
import { shallow } from 'enzyme';
import HomeContainer, { fetchData, addCoins } from "../HomeContainer";
// describe statement - Describes what is being tested
describe('HomeContainer', () => {
    const homeContainer = shallow(<HomeContainer />).dive();
// unit test, (what you are testing, testing a view)
    it('should render the HomeContainer without issues', () => {
        expect(homeContainer).toMatchSnapshot(); // Expect statement, what do you expect to happem. Snapshot - a picture of what your code looks like in code
    });

    describe('fetchdata()', () => {
        const url = '/someApi';
        const setStateData = jest.fn(); // Fake function to mock for test
        const setDataReloading = jest.fn();
        const setDataLoading = jest.fn();

        it('should set the reloading variable to true when init is true', () => {
            fetchData(true,url,setStateData,setDataReloading,setDataLoading)
            expect(setDataReloading).toHaveBeenCalledWith(true);
        });

        it('should set the data loading variable to true when init is false', () => {
            fetchData(false,url,setStateData,setDataReloading,setDataLoading)
            expect(setDataLoading).toHaveBeenCalledWith(true);
        });

        it('should call /someApi and then set the data state', () => {
            global.fetch = jest.fn(() =>
                Promise.resolve({
                    json: () => Promise.resolve([{name: 'R5', value: '5'}]),
                })
            );
            fetchData(true,url,setStateData,setDataReloading,setDataLoading)
            expect(fetch).toHaveBeenCalledWith(url);
            expect(setDataReloading).toHaveBeenCalledWith(true);
        });
    });

    describe('addCoins()', () => {
        const coinValue = 5;
        const setRunningTotal = jest.fn();
        const coins = [{name: 'R5', value: 5, amount: 4}, {name: 'R2', value: 2, amount: 2}];
        const runningTotal = 21;
        const setCoins = jest.fn();

        it('should set the running total and decrease the correct coins quantity', () => {
            addCoins(coinValue, setRunningTotal, coins, runningTotal, setCoins)
            expect(setRunningTotal).toHaveBeenCalledWith(26);
            expect(setCoins).toHaveBeenCalledWith([{"amount": 3, "name": "R5", "value": 5}, {"amount": 2, "name": "R2", "value": 2}]);
        });
    });
});
