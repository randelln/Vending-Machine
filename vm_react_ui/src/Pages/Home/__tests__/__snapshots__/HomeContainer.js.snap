// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeContainer should render the HomeContainer without issues 1`] = `
<Container
  className="main-container"
  style={
    Object {
      "paddingTop": "10px",
    }
  }
  tag="div"
>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <h3>
        Vending Machine
      </h3>
    </Col>
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <h3>
        Balance: R
        0
      </h3>
    </Col>
  </Row>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <DataTable
        fetchCoins={[Function]}
        fetchProducts={[Function]}
        loadingProducts={false}
        products={Array []}
        runningTotal={0}
        setLoadingBuy={[Function]}
        setRunningTotal={[Function]}
      />
    </Col>
  </Row>
  <tbody />
  <br />
  <tbody>
    <tr>
      <Button
        className="btn-padding-right"
        color="success"
        id="ReloadProductsButton"
        onClick={[Function]}
        tag="button"
      >
        Refresh products
      </Button>
      <Button
        className="btn-padding-right"
        color="success"
        id="ReloadCoinsButton"
        onClick={[Function]}
        tag="button"
      >
        Reload coins
      </Button>
      <Button
        className="btn-padding-right"
        color="success"
        id="ReloadAll"
        onClick={[Function]}
        tag="button"
      >
        Refresh All
      </Button>
      <Button
        className="btn-padding-right"
        color="danger"
        id="CancelButton"
        onClick={[Function]}
        tag="button"
      >
        Cancel
      </Button>
    </tr>
  </tbody>
</Container>
`;
